Hide empty tree items by default

iafan opened the pull request:
Empty folders or files with zero total units are now hidden by default. When displayed, they are rendered differently: as semitransparent labels, and with no progress bars. "Show/hide disabled" toggle now has an explanatory tooltip.

Also, when at any level all browser rows are disabled, we will show disabled items by default, saving admins a click and avoiding confusion when navigating disabled projects or empty folders.

julen commented on pootle/static/js/browser/components/BrowserTableRow.js:
```diff
@@ -81,6 +82,13 @@ const BrowserTableRow = React.createClass({
     if (itemType === ITEM_PROJECT) itemTypeName = 'project';
     if (itemType === ITEM_LANGUAGE) itemTypeName = 'language';
 
+    const progressBar = total === 0 ? '' :
```
Unless this is intentional, note that in order to display _nothing_, `null`, `false` or `undefined` should be used instead — `''` will create an unnecessary span/text elements otherwise. See http://stackoverflow.com/questions/30097091/correct-way-to-define-an-empty-dom-element-in-react for more details.

julen commented on pootle/static/js/browser/components/BrowserTable.js:
```diff
@@ -131,10 +131,16 @@ const BrowserTable = React.createClass({
     this.setState({ reverseSort, sortColumn });
   },
 
+  isItemDisabled(item) {
```
Nit: This doesn't depend on anything related to the class and personally I'd make it a regular function. However, use your best judgment.

julen commented on pootle/static/js/browser/components/BrowserTable.js:
```diff
@@ -131,10 +131,16 @@ const BrowserTable = React.createClass({
   },
 
+  isItemDisabled(item) {
+    return item.hasOwnProperty('is_disabled') && item.is_disabled
```
Nit: when splitting long lines, I think it'll be preferable to wrap everything in parens (we do that elsewhere AFAICT). Under some circumstance one might be tempted to wrap a long line like this:
```js
return
  imagine.this.is.a.very.very.long.line.with.a.stupid.methoCall(andMultiple, Arguments);
```
And unwillingly introduce a bug, because the function will always return `undefined` due to JS's Automatic Semicolon Insertion (ASI).

iafan commented:
I addressed all the nits.

> non-admin users are getting the toggle when they shouldn't

Good catch. I believe we don't want to send that data to non-admins at all, and also want to show 404 if non-admins want to go to the empty project/tp/directory/file. I added another commit to this PR that does just that. It comes with quite a few expected changes to snapshot files, and also I decided to store response code as snapshots since now tests can get either `200` or `404`.

julen commented:
> It comes with quite a few expected changes to snapshot files, and also I decided to store response code as snapshots since now tests can get either 200 or 404.

👍 for storing the response codes as they might differ.

Unless I am misreading, not all the changes are expected though: note how all non-admins get 404, but the test stores we have are not empty. I have the feeling this is because the stats cache is empty when running the tests. Unfortunately this doesn't have a straightforward solution I can think of right now without calculating stats, because the cache is the data store. So this is not like having a caching layer that would fall back to retrieve the data from the persistence layer.

iafan commented:
> not all the changes are expected though: note how all non-admins get 404, but the test stores we have are not empty

Yep. Children don't have stats, which is the same as having zero totals, and this is why we _expect_ the view to return 404. However, having no stats in test context data is definitely _not expected_, so let's think how this can be addressed.

julen commented:
> However, having no stats in test context data is definitely not expected, so let's think how this can be addressed.

Today I have been fighting to make stats available to tests, please check #138 for further details.

iafan commented:
@julen I have rebased PR against master and updated snapshots. See the diff — something's still not good with the snapshots, as language views don't have stats and thus return 404. Other views seem to work fine now.

julen commented:
This outlines a problem originating in `TreeItem`. Note this class holds a collection of children references, and provides aggregated values which are calculated _live_ out of existing cached children stats. When one of the children elements is empty, e.g. a translation project containing no stores or units, it will still try to retrieve its stats, and failure to do so (signaled via the `NoCachedStats` exception) will basically abort further data aggregation, therefore returning default values.

In tests this happens because a `terminology` project is created and then it is assigned to all test languages. The resulting terminology TPs have no stores/units at all (can't tell what's the purpose of it TBH), and hence no cached stats data.

I am not entirely sure whether we allow to have empty TPs at all and this situation is realistic (although it's late here so I want to stress again I am not sure). In case empty TPs become obsolete, they would never be part of `TreeItem`'s children, and there would be no chance of hitting this problem. If there is a remote chance this situation can realistically happen, then we are in front of a bug in `TreeItem`.

iafan commented:
Languages and TPs are unlikey to become empty, but wouldn't the same issue happen on e.g. empty folder (this is a pretty common thing)?

julen commented:
No, this only affects language children (and _maybe_ cross-language resource views). This is because they use the `TreeItem` class, as opposed to `CachedTreeItem`. [As commented before](https://github.com/evernote/zing/pull/129#issuecomment-268399281), the difference is notable: the former aggregates at run-time the values which have already been calculated in children elements; the latter keeps track of calculation results in the stats cache. If a folder is empty, its cache key already reports empty values (otherwise it will be dirty).

A quick fix is to move the exception handling one level below, and that would just work. However, if the situation is not realistic or possible in a production server, then the test environment setup needs to be adjusted instead of application code, so that the active TPs have at least a unit.

julen commented:
I was going through `TreeItem` code to allow returning some stats, although I realize now this behavior has been intentional, because the old implementation of the table was able to differentiate between the situation where there are no stats (empty cache) and stats with zero values.

We must as well note that since we re-implemented the browsing table, in the UI now we don't make this distinction. If we apply this PR, this means that if the stats cache is empty for any tree item, users will get 404 everywhere where we previously (in the old implementation) would see `—` values. Admins can still access browsing pages, but they will see all items as empty.

So I believe we will need to step back a bit here and think about what we want to do with empty vs. zero values. My feeling is we will want to keep this distinction, although note that we now remove falsy values in the backend, and the UI has no notion of uninitialized values. Hence we will need to ensure first the browsing table is capable of displaying meaningful information under these conditions.

iafan commented:
Can we indicate 'unknown' stats by returning e.g. `-1` as a total value?

julen commented:
> I am not entirely sure whether we allow to have empty TPs at all and this situation is realistic or not

After giving this another thought, I am sure this is not realistic.

Notice how the test setup is incomplete: it manually creates the _terminology_ project and adds languages (TPs) to it, without setting up any stores. In real life, the only way to initialize a TP is by either a) running `update_stores` with files on disk, or b) adding a new project from the admin UI, which will scan files on disk and initialize TPs. In this scenario, if the TP directory contains no files/units, the TP directory will be marked as obsolete. This means the stats code won't go through it when aggregating children values.

For the sake of this PR I am going to fix the test environment. We will still need to figure out how we deal with the distinction between missing stats and zero values, but let's keep that separate and not block the work here on that.

julen commented:
The environment setup has been fixed in #150, so that terminology projects have at least a store with some units.

On the other hand, in #151 I added two types of _empty_ directories which are required for this PR: a directory with no contents at all (marked as obsolete under these circumstances), and another directory with a store that has no units. Based on that data, I am positive this PR is unlocked.

By the way, and related to the previous paragraph, I wonder what's the motivation we have for allowing the second type of empty directories in the first place. I am not entirely sure if I am missing something, but if the store is empty, it serves no purpose. If we would ignore/obsolete such stores (plus directories), then we wouldn't even need this PR at all, because when querying the data the browser code would only return children items which are alive.

iafan commented:
> I am not entirely sure if I am missing something, but if the store is empty, it serves no purpose. If we would ignore/obsolete such stores (plus directories), then we wouldn't even need this PR at all, because when querying the data the browser code would only return children items which are alive.

If we have an empty .po file for some reason, it makes sense to display it for admins as an empty store — this will help avoid confusion when debugging the whole sync process. Otherwise the only way to understand why some file is not appearing in the UI is to SSH to the server, find relevant .po files and see if they are empty.

The way system behaves currently is more explicit and transparent, which I like. Ultimately, if one doesn't want empty stores to be present in Zing, they shouldn't be creating empty translation files in the first place, instead of relying on Zing to silently 'fix' this at update_stores time.

iafan commented:
@julen I rebased the PR on top of the recent changes in master, please verify the changes in snapshots once again. From what I see, the changes seem to be legit now.

julen commented:
The snapshots look fine now although you will need to rebase the PR again.
